kaspad/src/daemon.rs

```diff
@@ -0,0 +1,357 @@
+use kaspa_mining::manager::{MiningManager, MiningManagerProxy};
+use kaspa_p2p_flows::{flow_context::FlowContext, service::P2pService};
+
+use kaspa_perf_monitor::builder::Builder as PerfMonitorBuilder;
+use kaspa_utxoindex::{api::UtxoIndexProxy, UtxoIndex};
+use kaspa_wrpc_server::service::{Options as WrpcServerOptions, ServerCounters as WrpcServerCounters, WrpcEncoding, WrpcService};
+
+use crate::args::Args;
+
+const DEFAULT_DATA_DIR: &str = "datadir";
+const CONSENSUS_DB: &str = "consensus";
+const UTXOINDEX_DB: &str = "utxoindex";
+const META_DB: &str = "meta";
+const DEFAULT_LOG_DIR: &str = "logs";
+
+fn get_home_dir() -> PathBuf {
+    #[cfg(target_os = "windows")]
+    return dirs::data_local_dir().unwrap();
+    #[cfg(not(target_os = "windows"))]
+    return dirs::home_dir().unwrap();
+}
+
+fn get_app_dir() -> PathBuf {
+    #[cfg(target_os = "windows")]
+    return get_home_dir().join("rusty-kaspa");
+    #[cfg(not(target_os = "windows"))]
+    return get_home_dir().join(".rusty-kaspa");
+}
+
+fn validate_config_and_args(_config: &Arc<Config>, args: &Args) -> ConfigResult<()> {
+    if !args.connect_peers.is_empty() && !args.add_peers.is_empty() {
+        return Err(ConfigError::MixedConnectAndAddPeers);
+    }
+    if args.logdir.is_some() && args.no_log_files {
+        return Err(ConfigError::MixedLogDirAndNoLogFiles);
+    }
+    Ok(())
+}
+
+fn get_user_approval_or_exit(message: &str, approve: bool) {
+    if approve {
+        return;
+    }
+    println!("{}", message);
+    let mut input = String::new();
+    match std::io::stdin().read_line(&mut input) {
+        Ok(_) => {
+            let lower = input.to_lowercase();
+            let answer = lower.as_str().strip_suffix("\r\n").or(lower.as_str().strip_suffix('\n')).unwrap_or(lower.as_str());
+            if answer == "y" || answer == "yes" {
+                // return
+            } else {
+                println!("Operation was rejected ({}), exiting..", answer);
+                exit(1);
+            }
+        }
+        Err(error) => {
+            println!("Error reading from console: {error}, exiting..");
+            exit(1);
+        }
+    }
+}
+
+#[derive(Default)]
+pub struct Runtime {
+    log_dir: Option<String>,
+}
+
+impl Runtime {
+    pub fn from_args(args: &Args) -> Self {
+        // Configure the panic behavior
+        kaspa_core::panic::configure_panic();
+
+        let network = args.network();
+
+        let config = Arc::new(
+            ConfigBuilder::new(network.into())
+                .adjust_perf_params_to_consensus_params()
+                .apply_args(|config| args.apply_to_config(config))
+                .build(),
+        );
+
+        // Make sure config and args form a valid set of properties
+        if let Err(err) = validate_config_and_args(&config, args) {
+            println!("{}", err);
+            exit(1);
+        }
+
+        // TODO: Refactor all this quick-and-dirty code
+        let app_dir = args
```

Comment:
There's an identical code block below at `create_core_with_runtime`. Depending on the intended usage we can either store `app_dir` within `Runtime` and reuse it below, or we can extract the logic to a function and call from both locations (in case we want `app_dir` to be different between core instances). I would also remove the `// TODO: Refactor all this quick-and-dirty code`